internal/services/network/custom_ip_prefix_resource.go

```diff
@@ -0,0 +1,419 @@
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/network/parse"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/network/validate"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tags"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/validation"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/timeouts"
+	"github.com/hashicorp/terraform-provider-azurerm/utils"
+)
+
+type commissionedAction string
+
+const (
+	commissionedActionProvision    commissionedAction = "Provision"
+	commissionedActionCommission   commissionedAction = "Commission"
+	commissionedActionDecommission commissionedAction = "Decommission"
+	commissionedActionDeprovision  commissionedAction = "Deprovision"
+)
+
+type commissionedState string
+
+const (
+	commissionedStateDeprovisioned    commissionedState = "Deprovisioned"
+	commissionedStateValidationFailed commissionedState = "ValidationFailed"
+)
+
+func resourceCustomIpPrefix() *pluginsdk.Resource {
+	return &pluginsdk.Resource{
+		Create: resourceCustomIpPrefixCreateUpdate,
+		Read:   resourceCustomIpPrefixRead,
+		Update: resourceCustomIpPrefixCreateUpdate,
+		Delete: resourceCustomIpPrefixDelete,
+
+		Importer: pluginsdk.ImporterValidatingResourceId(func(id string) error {
+			_, err := parse.CustomIpPrefixID(id)
+			return err
+		}),
+
+		Timeouts: &pluginsdk.ResourceTimeout{
+			Create: pluginsdk.DefaultTimeout(30 * time.Minute),
+			Read:   pluginsdk.DefaultTimeout(5 * time.Minute),
+			Update: pluginsdk.DefaultTimeout(30 * time.Minute),
+			Delete: pluginsdk.DefaultTimeout(30 * time.Minute),
+		},
+
+		Schema: map[string]*pluginsdk.Schema{
+			"name": {
+				Type:         pluginsdk.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.CustomIpPrefixName,
+			},
+
+			"location": commonschema.Location(),
+
+			"resource_group_name": commonschema.ResourceGroupName(),
+
+			"cidr": {
+				Type:         pluginsdk.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.IsCIDR,
+			},
+
+			"zones": commonschema.ZonesMultipleRequiredForceNew(),
+
+			"authorization_message": {
+				Type:         pluginsdk.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"signed_message": {
+				Type:         pluginsdk.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"action": {
+				Type:     pluginsdk.TypeString,
+				Optional: true,
+				Default:  string(commissionedActionProvision),
+				ValidateFunc: validation.StringInSlice([]string{
+					string(commissionedActionProvision),
+					string(commissionedActionCommission),
+					string(commissionedActionDecommission),
+					string(commissionedActionDeprovision),
+				}, false),
+			},
```

Comment:
why is this exposed? isn't this the internal state?

Reply:
The commissionedState is not internal state since it could be updated after the custom ip prefix resource is created successfully.